File: java/leetcode_grind/Day843.java
package leetcode_grind;

public class Day843 {
    // https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/description/?envType=daily-question&envId=2025-03-11
    static class Solution1 {
        public int numberOfSubstrings(String s) {
            int len = s.length();
            int left = 0, right = 0;
            int[] freq = new int[3];
            int total = 0;

            while (right < len) {
                char curr = s.charAt(right);
                freq[curr - 'a']++;

                while (hasAllChars(freq)) {
                    total += len - right;
                    char leftChar = s.charAt(left);
                    freq[leftChar - 'a']--;
                    left++;
                }

                right++;
            }

            return total;
        }

        boolean hasAllChars(int[] freq) {
            return freq[0] > 0 && freq[1] > 0 && freq[2] > 0;
        }
    }

}

File: src/leetcode_grind/day_843.rs
// https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/description/?envType=daily-question&envId=2025-03-11
pub fn number_of_substrings(s: String) -> i32 {
    let s = s.chars().collect::<Vec<_>>();
    let len = s.len();        
    let mut last_pos = vec![-1,-1,-1];
    let mut total = 0;
    for pos in 0..len {
        last_pos[s[pos] as usize - 'a' as usize] = pos as i32;
        total += 1 + last_pos[0].min(last_pos[1]).min(last_pos[2]);
    }
    total as i32
}